src/Database/LSMTree/Internal/BlobRef.hs

```diff
@@ -48,7 +60,56 @@ instance NFData BlobSpan where
 blobRefSpanSize :: BlobRef m h -> Int
 blobRefSpanSize = fromIntegral . blobSpanSize . blobRefSpan
 
--- | The 'BlobSpan' to read must come from this run!
+-- | A 'WeakBlobRef' is a weak reference to a blob file. These are the ones we
+-- can return in the public API and can outlive their parent table. They do not
+-- keep the file open using a reference count. So when we want to use our weak
+-- reference we have to dereference them to obtain a normal strong reference
+-- while we do the I\/O to read the blob. This ensures the file is not closed
+-- under our feet.
+--
+newtype WeakBlobRef m h = WeakBlobRef (BlobRef m h)
+  deriving newtype (Eq, Show, NFData)
+
+newtype DeRefWeakBlobRef = DeRefWeakBlobRef Int
```

Comment:
It wasn't immediately clear to me what this exception means -- maybe the name should more closely resemble what goes wrong that leads to this exception

Reply:
especially as this exception will be thrown to the users, AFAICT.

Reply:
No, it's internal. It gets caught and re-thrown using the existing exception type in the public API.

Reply:
I'll rename it `WeakBlobRefInvalid`.

Reply:
I'm not a fan of rethrowing. What prevents us from throwing the public exception immediately (and not kind of using exceptions for control flow)?